File: bridge/test/fake-gather.js
/**
 * A fake Gather game server, shared by `direct.test.js` and `bridge.test.js`.
 *
 * Not a `*.test.js` file, so `npm test` does not try to run it.
 *
 * Test-only, like the encoder in `msgpack.test.js`. `WsConnection` only surfaces
 * *text* frames — it exists to serve JSON to phones — and the game protocol is
 * binary, so the framing here is hand-rolled rather than reused.
 *
 * It speaks just enough of the real protocol to be worth testing against: the
 * handshake is recorded verbatim so assertions can check what the collector sent,
 * and state is only dumped for a connection that actually authenticated. Ids and
 * names are synthetic; the op names, envelope keys and field spellings are real.
 */

import { createHash } from 'node:crypto';
import { createServer } from 'node:http';

import { decode, encode } from '../lib/msgpack.js';
import { encodeFrame } from '../lib/ws.js';

/** The default dump: me, and one person standing one tile away. */
export function defaultPatches() {
  return [
    {
      op: 'addmodel',
      model: 'Connection',
      data: {
        id: 'conn-1',
        spaceId: 'space-1',
        authUserId: 'uid-1',
        spaceUserId: 'me-1',
        entered: false,
        target: 'OfficeView',
      },
    },
    { op: 'addmodel', model: 'Space', data: { id: 'space-1', name: 'Test Space' } },
    {
      op: 'addmodel',
      model: 'SpaceUser',
      data: {
        id: 'me-1',
        name: 'Me',
        position: { $type: 'Position', x: 10, y: 10 },
        floorId: 'floor-1',
        connected: true,
        isBot: false,
      },
    },
    {
      op: 'addmodel',
      model: 'SpaceUser',
      data: {
        id: 'them-1',
        name: 'Neighbour',
        position: { $type: 'Position', x: 11, y: 10 },
        floorId: 'floor-1',
        connected: true,
        isBot: false,
      },
    },
  ];
}

export function fakeGameServer({ requireAuth = true, patches = defaultPatches } = {}) {
  const connections = [];
  const server = createServer();

  server.on('upgrade', (req, socket) => {
    const key = req.headers['sec-websocket-key'];
    const accept = createHash('sha1')
      .update(`${key}258EAFA5-E914-47DA-95CA-C5AB0DC85B11`)
      .digest('base64');
    socket.write(
      'HTTP/1.1 101 Switching Protocols\r\n' +
        'Upgrade: websocket\r\n' +
        'Connection: Upgrade\r\n' +
        `Sec-WebSocket-Accept: ${accept}\r\n\r\n`,
    );

    const send = (frame) => socket.write(encodeFrame(2, encode(frame)));
    const conn = { url: req.url, received: [], authenticated: false, socket, send };
    connections.push(conn);

    /** Push a delta after the dump, the way the real server does. */
    conn.delta = (list) => send({ type: 'DeltaState', sequenceNumber: 2, patches: list });

    /**
     * Push interaction events onto Gather's event bus.
     *
     * A real wave arrives in a `DeltaState` whose `patches` array is **empty** —
     * which is precisely why the bus went unread for so long, and so exactly what
     * a test needs to reproduce.
     */
    conn.bus = (events) =>
      send({ type: 'DeltaState', sequenceNumber: 3, patches: [], actionReturns: [], events });

    // The real server heartbeats before the first state chunk, which is what
    // makes a premature "handshake rejected" verdict so tempting.
    send({ type: 'Heartbeat', timestamp: 1, origin: 'Server' });

    let buf = Buffer.alloc(0);
    socket.on('data', (chunk) => {
      buf = Buffer.concat([buf, chunk]);
      for (;;) {
        const frame = readClientFrame(buf);
        if (!frame) break;
        buf = frame.rest;
        if (frame.opcode >= 0x8) continue; // control frames: ignore
        let decoded;
        try {
          decoded = decode(frame.payload);
        } catch {
          continue;
        }
        conn.received.push(decoded);

        if (decoded?.type === 'Authenticate') {
          conn.authenticated = decoded?.credential?.type === 'JWT' && !!decoded?.credential?.jwt;
        }
        if (decoded?.action === 'loadSpaceUser') {
          if (requireAuth && !conn.authenticated) continue; // stay silent, as Gather does
          send({ type: 'SpaceStatus', warmInGatewayServer: true });
          send({ type: 'FullStateChunk', sequenceNumber: 1, fullStatePatches: patches() });
          conn.dumped = true;
        }
      }
    });
    socket.on('error', () => {});
  });

  return {
    connections,
    /** The most recent connection, which is the one a test just triggered. */
    get latest() {
      return connections[connections.length - 1] ?? null;
    },
    listen: () =>
      new Promise((resolve) => {
        server.listen(0, '127.0.0.1', () => resolve(`ws://127.0.0.1:${server.address().port}`));
      }),
    close: () => {
      for (const c of connections) {
        try {
          c.socket.destroy();
        } catch {
          /* gone */
        }
      }
      return new Promise((resolve) => server.close(resolve));
    },
  };
}

/** A `MeetingParticipant` row: an invite when `inviterId` is set. */
export function participant({
  id = 'part-1',
  spaceUserId = 'me-1',
  inviterId = null,
  meetingId = 'meeting-1',
} = {}) {
  return {
    op: 'addmodel',
    model: 'MeetingParticipant',
    data: {
      id,
      spaceUserId,
      meetingId,
      ...(inviterId ? { inviterId } : {}),
      inviteStatus: inviterId ? 'InvitedRequired' : 'NotInvited',
      createdAt: '2026-08-07T08:01:42.405Z',
    },
  };
}

/** A `MeetingJoinRequest`: somebody knocking, unanswered unless `respondedAt`. */
export function joinRequest({
  id = 'join-1',
  spaceUserId = 'them-1',
  meetingId = 'meeting-1',
  respondedAt = null,
} = {}) {
  return {
    op: 'addmodel',
    model: 'MeetingJoinRequest',
    data: {
      id,
      spaceUserId,
      meetingId,
      ...(respondedAt ? { respondedAt } : {}),
      createdAt: '2026-08-07T09:45:21.834Z',
    },
  };
}

/**
 * One `WaveEvent`, shaped exactly as captured from a live space on 2026-08-07.
 *
 * The two halves live in different places on purpose, because they do in the real
 * envelope: who waved is `payload.senderId`, and who they waved *at* is
 * `options.targetUserIds`.
 */
export function waveEvent({
  senderId = 'them-1',
  targetId = 'me-1',
  sentTime = '2026-08-07T14:22:20.563Z',
} = {}) {
  return {
    payload: { eventName: 'WaveEvent', senderId, sentTime },
    options: { targetUserIds: [targetId] },
  };
}

/**
 * A syntactically valid JWT carrying a chosen uid. Collectors read the uid
 * straight out of the token, so this also keeps tests from falling back to
 * whatever session the developer happens to have adopted.
 */
export function fakeJwt(uid = 'uid-1') {
  const claims = Buffer.from(JSON.stringify({ user_id: uid, exp: 4e9 })).toString('base64url');
  return `eyJhbGciOiJub25lIn0.${claims}.sig`;
}

/** Minimal masked-frame reader; returns null until a whole frame is buffered. */
function readClientFrame(buf) {
  if (buf.length < 2) return null;
  const opcode = buf[0] & 0x0f;
  let length = buf[1] & 0x7f;
  let offset = 2;
  if (length === 126) {
    if (buf.length < 4) return null;
    length = buf.readUInt16BE(2);
    offset = 4;
  } else if (length === 127) {
    if (buf.length < 10) return null;
    length = Number(buf.readBigUInt64BE(2));
    offset = 10;
  }
  if (buf.length < offset + 4 + length) return null;
  const mask = buf.subarray(offset, offset + 4);
  offset += 4;
  const payload = Buffer.from(buf.subarray(offset, offset + length));
  for (let i = 0; i < payload.length; i++) payload[i] ^= mask[i & 3];
  return { opcode, payload, rest: buf.subarray(offset + length) };
}

File: packages/gather_client/test/fake_gather.dart
/// A fake Gather game server, the Dart counterpart of `bridge/test/fake-gather.js`.
///
/// A real `HttpServer` doing a real WebSocket upgrade, so [DirectCollector] runs its
/// production socket path — `WebSocket.connect`, binary frames, close codes and all.
/// Mocking the socket would leave the one part most likely to be wrong untested.
///
/// It speaks just enough of the protocol to be worth testing against: the handshake
/// is recorded verbatim so assertions can check what the collector sent, and state
/// is only dumped for a connection that actually authenticated — because Gather's
/// distinguishing behaviour is that it stays *silent* on a bad handshake rather
/// than rejecting it.
library;

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:gather_client/gather_client.dart';

/// The default dump: me, and one other person.
List<Map<String, Object?>> defaultPatches() => [
      {
        'op': 'addmodel',
        'model': 'Connection',
        'data': {
          'id': 'conn-1',
          'spaceId': 'space-1',
          'authUserId': 'uid-1',
          'spaceUserId': 'me-1',
          'entered': false,
          'target': 'OfficeView',
        },
      },
      {
        'op': 'addmodel',
        'model': 'Space',
        'data': {'id': 'space-1', 'name': 'Test Space'},
      },
      {
        'op': 'addmodel',
        'model': 'SpaceUser',
        'data': {
          'id': 'me-1',
          'name': 'Me',
          'position': {r'$type': 'Position', 'x': 10, 'y': 10},
          'floorId': 'floor-1',
          'connected': true,
          'isBot': false,
        },
      },
      {
        'op': 'addmodel',
        'model': 'SpaceUser',
        'data': {
          'id': 'them-1',
          'name': 'Neighbour',
          'position': {r'$type': 'Position', 'x': 11, 'y': 10},
          'floorId': 'floor-1',
          'connected': true,
          'isBot': false,
        },
      },
    ];

/// One `WaveEvent`, shaped exactly as captured from a live space on 2026-08-07.
///
/// The two halves live in different places on purpose, because they do in the real
/// envelope: who waved is `payload.senderId`, who they waved *at* is
/// `options.targetUserIds`.
Map<String, Object?> waveEvent({
  String senderId = 'them-1',
  String targetId = 'me-1',
  String sentTime = '2026-08-07T14:22:20.563Z',
}) =>
    {
      'payload': {'eventName': 'WaveEvent', 'senderId': senderId, 'sentTime': sentTime},
      'options': {
        'targetUserIds': [targetId],
      },
    };

class FakeConnection {
  FakeConnection(this.socket, this.url);

  final WebSocket socket;
  final String url;

  /// Every frame the client sent, decoded.
  final List<Map<String, Object?>> received = [];
  bool authenticated = false;
  bool dumped = false;

  void send(Map<String, Object?> frame) {
    if (socket.readyState == WebSocket.open) socket.add(msgpackEncode(frame));
  }

  /// Push a delta after the dump, the way the real server does.
  void delta(List<Map<String, Object?>> patches) =>
      send({'type': 'DeltaState', 'sequenceNumber': 2, 'patches': patches});

  /// Push interaction events onto Gather's event bus.
  ///
  /// A real wave arrives in a `DeltaState` whose `patches` array is **empty** —
  /// which is why the bus went unread for so long, and so exactly what a test needs.
  void bus(List<Map<String, Object?>> events) => send({
        'type': 'DeltaState',
        'sequenceNumber': 3,
        'patches': <Object?>[],
        'actionReturns': <Object?>[],
        'events': events,
      });

  /// Refuse the last action of [action], the way the real gateway does.
  ///
  /// The refusal is addressed by `txnId`, so the transaction has to be looked up in
  /// what the client actually sent — which is the point: pairing an answer back to
  /// the question is the whole job of the code under test. Answers arrive with an
  /// **empty** `patches` array, because a refused action changes nothing.
  void refuse(String action, Object? error) {
    final sent = received.lastWhere(
      (frame) => frame['action'] == action,
      orElse: () => throw StateError('the client never sent $action'),
    );
    send({
      'type': 'DeltaState',
      'sequenceNumber': 4,
      'patches': <Object?>[],
      'actionReturns': [
        {
          'connectionId': 'conn-1',
          'txnId': sent['txnId'],
          'result': {'type': 'Error', 'error': error},
        },
      ],
    });
  }
}

class FakeGatherServer {
  FakeGatherServer._(this._server, this.url);

  final HttpServer _server;

  /// `ws://127.0.0.1:<port>` — pass straight to [DirectCollector.socketUrl].
  final String url;

  final List<FakeConnection> connections = [];
  final _connected = StreamController<FakeConnection>.broadcast();

  /// Fires as each client finishes its handshake and gets its dump.
  Stream<FakeConnection> get onDumped => _connected.stream;

  FakeConnection? get latest => connections.isEmpty ? null : connections.last;

  static Future<FakeGatherServer> start({
    bool requireAuth = true,
    List<Map<String, Object?>> Function()? patches,
  }) async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    final fake = FakeGatherServer._(server, 'ws://127.0.0.1:${server.port}');

    server.listen((request) async {
      final socket = await WebSocketTransformer.upgrade(request);
      final conn = FakeConnection(socket, request.uri.toString());
      fake.connections.add(conn);

      // The real server heartbeats before the first state chunk, which is what makes
      // a premature "handshake rejected" verdict so tempting.
      conn.send({'type': 'Heartbeat', 'timestamp': 1, 'origin': 'Server'});

      socket.listen((data) {
        if (data is! List<int>) return;
        final Object? decoded;
        try {
          decoded = msgpackDecode(Uint8List.fromList(data));
        } on Object {
          return;
        }
        if (decoded is! Map<String, Object?>) return;
        conn.received.add(decoded);

        if (decoded['type'] == 'Authenticate') {
          final credential = decoded['credential'];
          conn.authenticated = credential is Map &&
              credential['type'] == 'JWT' &&
              (credential['jwt'] as String?)?.isNotEmpty == true;
        }
        if (decoded['action'] == 'loadSpaceUser') {
          // Stay silent when unauthenticated, as Gather does.
          if (requireAuth && !conn.authenticated) return;
          conn.send({'type': 'SpaceStatus', 'warmInGatewayServer': true});
          conn.send({
            'type': 'FullStateChunk',
            'sequenceNumber': 1,
            'fullStatePatches': (patches ?? defaultPatches)(),
          });
          conn.dumped = true;
          if (!fake._connected.isClosed) fake._connected.add(conn);
        }
      }, onError: (Object _) {});
    });

    return fake;
  }

  Future<void> close() async {
    for (final c in connections) {
      try {
        await c.socket.close();
      } on Object {
        /* gone */
      }
    }
    await _connected.close();
    await _server.close(force: true);
  }
}

/// A syntactically valid JWT carrying a chosen uid.
///
/// The collector reads the uid straight out of the token, so this also keeps tests
/// from depending on whatever session the developer happens to have paired.
String fakeJwt({String uid = 'uid-1', Duration validFor = const Duration(hours: 1)}) {
  final exp = DateTime.now().add(validFor).millisecondsSinceEpoch ~/ 1000;
  final claims = _b64({'user_id': uid, 'exp': exp});
  return 'eyJhbGciOiJub25lIn0.$claims.sig';
}

String _b64(Map<String, Object?> claims) {
  final json = claims.entries
      .map((e) => '"${e.key}":${e.value is String ? '"${e.value}"' : e.value}')
      .join(',');
  return base64UrlNoPad('{$json}');
}

String base64UrlNoPad(String text) {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
  final bytes = text.codeUnits;
  final out = StringBuffer();
  for (var i = 0; i < bytes.length; i += 3) {
    final b0 = bytes[i];
    final b1 = i + 1 < bytes.length ? bytes[i + 1] : null;
    final b2 = i + 2 < bytes.length ? bytes[i + 2] : null;
    out.write(chars[b0 >> 2]);
    out.write(chars[((b0 & 0x03) << 4) | ((b1 ?? 0) >> 4)]);
    if (b1 != null) out.write(chars[((b1 & 0x0f) << 2) | ((b2 ?? 0) >> 6)]);
    if (b2 != null) out.write(chars[b2 & 0x3f]);
  }
  return out.toString();
}

/// A [GatherHttp] that never touches the network.
///
/// A suite that reached Google would be slow, flaky, and would burn the developer's
/// real session.
class FakeGatherHttp implements GatherHttp {
  FakeGatherHttp({
    this.idToken,
    this.status = 200,
    this.errorCode,
    this.spaces = const {},
  });

  String? idToken;
  int status;
  String? errorCode;
  Map<String, Object?> spaces;

  int refreshes = 0;

  @override
  Future<({int status, Map<String, Object?> body})> postForm(
    Uri uri,
    Map<String, String> fields,
  ) async {
    refreshes++;
    if (status != 200) {
      return (
        status: status,
        body: {
          'error': {'message': errorCode ?? 'SOMETHING_ELSE'},
        },
      );
    }
    return (
      status: 200,
      body: {
        'id_token': idToken ?? fakeJwt(),
        'refresh_token': fields['refresh_token'],
        'user_id': 'uid-1',
      },
    );
  }

  @override
  Future<({int status, Object? body})> getJson(Uri uri, String bearer) async =>
      (status: 200, body: spaces);

  /// Msgpack bodies, keyed by the last path segment of the URI they answer.
  Map<String, List<int>> bytes = const {};

  /// The status [getBytes] and [postBytes] answer with. Separate from [status],
  /// which belongs to the token endpoint — a suite needs to fail one without
  /// failing the other.
  int byteStatus = 200;

  /// Every [postJson] call, in order, so a test can assert what was sent.
  final List<({Uri uri, Object? body})> posts = [];

  @override
  Future<({int status, List<int> body})> getBytes(Uri uri, String bearer) async => (
        status: byteStatus,
        body: bytes[uri.pathSegments.last] ?? const <int>[],
      );

  @override
  Future<({int status, List<int> body})> postJson(
    Uri uri,
    String bearer,
    Object? body,
  ) async {
    posts.add((uri: uri, body: body));
    return (status: byteStatus, body: const <int>[]);
  }
}
